processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -127,45 +130,75 @@ protected Integer addToFacts(
       ThreadLocal<InputRow> in
   ) throws IndexSizeExceededException
   {
-    Integer rowOffset;
-    synchronized (this) {
-      rowOffset = numEntries.get();
-      if(rowOffset >= maxRowCount && !facts.containsKey(key)) {
+    final Integer priorIdex = facts.get(key);
+
+    Aggregator[] aggs;
+
+    if (null != priorIdex) {
+      aggs = concurrentGet(priorIdex);
+    } else {
+      aggs = new Aggregator[metrics.length];
+      for (int i = 0; i < metrics.length; i++) {
+        final AggregatorFactory agg = metrics[i];
+        aggs[i] = agg.factorize(
+            makeColumnSelectorFactory(agg, in, deserializeComplexMetrics)
+        );
+      }
+      final Integer rowIndex = indexIncrement.getAndIncrement();
+
+      concurrentSet(rowIndex, aggs);
+
+      // Last ditch sanity checks
+      if (numEntries.get() >= maxRowCount && !facts.containsKey(key)) {
         throw new IndexSizeExceededException("Maximum number of rows reached");
       }
-      final Integer prev = facts.putIfAbsent(key, rowOffset);
-      if (prev != null) {
-        rowOffset = prev;
-      } else {
-        Aggregator[] aggs = new Aggregator[metrics.length];
-        for (int i = 0; i < metrics.length; i++) {
-          final AggregatorFactory agg = metrics[i];
-          aggs[i] = agg.factorize(
-              makeColumnSelectorFactory(agg, in, deserializeComplexMetrics)
-          );
-        }
-        aggList.add(aggs);
+      final Integer prev = facts.putIfAbsent(key, rowIndex);
+      if (null == prev) {
         numEntries.incrementAndGet();
+      } else {
+        // We lost a race
+        aggs = concurrentGet(prev);
+        // Free up the misfire
+        concurrentRemove(rowIndex);
+        // This is expected to occur ~80% of the time in the worst scenarios
       }
     }
 
     in.set(row);
 
-    final Aggregator[] aggs = aggList.get(rowOffset);
-    for (int i = 0; i < aggs.length; i++) {
-      synchronized (aggs[i]) {
-        aggs[i].aggregate();
+    for (Aggregator agg : aggs) {
+      synchronized (agg) {
+        agg.aggregate();
       }
     }
+
     in.set(null);
+
+
     return numEntries.get();
   }
 
+  protected Aggregator[] concurrentGet(int offset)
```

Comment:
Easier readability and I'm working on updates to IncrementalIndex which uses them.